### ISSUE TITLE
[Bounty $1,000] ttnn.prod_bw returns non-finite gradients for zero inputs

### ISSUE THREAD

**Issue by kanapitsas** [tenstorrent/tt-metal#54551]
## Summary

`ttnn.prod_bw` computes the gradient as

```text
prod(input) * grad / input
```

which is `reciprocal(input)` multiplied by a broadcast `prod(input) * grad`. Where the input
contains an exact zero, `reciprocal(0)` is `+inf` and the surviving product is `0`, so the
gradient at that position is non-finite instead of the finite value autograd defines.

```text
input:    [2, 0, 4]
grad:     1
expected: [0, 8, 0]
current:  [0, non-finite, 0]
```

This is a source-level analysis: I do not have a card, so the host replay below renders the
device `inf * 0` as a non-finite value without asserting which one appears in silicon. The defect
is that the gradient is non-finite where the correct value is finite.

## Root cause

Both branches multiply by `reciprocal(input)`:

- `dim=None`:
  [`unary_backward.cpp#L1668-L1676`](https://github.com/tenstorrent/tt-metal/blob/a5cd86212e/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L1668-L1676)
  — `multiply(reciprocal(input), fill_first_val_into_tensor(prod * grad))`.
- per-dimension: `reciprocal_input` is built once at
  [`L1711`](https://github.com/tenstorrent/tt-metal/blob/a5cd86212e/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L1711)
  and feeds every `dim` branch, e.g. `dim=-1` at
  [`L1722`](https://github.com/tenstorrent/tt-metal/blob/a5cd86212e/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L1722).
  All four dimensions are affected, not only `dim=-1`.

At a zero input the product of the other values has already been lost in the reduction, so it
cannot be recovered by the division. The formula is valid only when the reduced input has no zeros.

## Reproducer

Host only, no device required (`pip install torch`):

```python
"""Source-level replay of ttnn.prod_bw against torch.prod autograd.

Replays the composite expression from tt-metal main @ a5cd86212e. Runs on host only:
it does not need a Tenstorrent device.
"""

import math
import torch


def source_prod_bw(values, grad):
    """unary_backward.cpp L1668-1676: reciprocal(input) * broadcast(prod(input) * grad)."""
    product = math.prod(values)
    out = []
    for v in values:
        if v == 0.0:
            # reciprocal(0) is +inf on device; inf * 0 is non-finite
            out.append(float("inf") * (product * grad))
        else:
            out.append(product * grad / v)
    return out


def source_prod_bw_dim(rows, grad):
    """unary_backward.cpp L1711-1722, dim=-1: bcast(reciprocal(input), prod * grad, MUL)."""
    return [source_prod_bw(row, grad) for row in rows]


def torch_prod_bw(values, grad):
    t = torch.tensor(values, dtype=torch.float64, requires_grad=True)
    torch.prod(t).backward(torch.tensor(grad, dtype=torch.float64))
    return t.grad.tolist()


def torch_prod_bw_dim(rows, grad):
    t = torch.tensor(rows, dtype=torch.float64, requires_grad=True)
    torch.prod(t, dim=-1, keepdim=True).backward(
        torch.full((len(rows), 1), grad, dtype=torch.float64))
    return t.grad.tolist()


def show(v):
    return "non-finite" if isinstance(v, float) and not math.isfinite(v) else f"{v:g}"


print("dim=None")
for values in ([2.0, 3.0, 4.0], [2.0, 0.0, 4.0], [0.0, 0.0, 4.0], [1.0, 5.0, 0.0]):
    actual, expected = source_prod_bw(values, 1.0), torch_prod_bw(values, 1.0)
    ok = all(a == e for a, e in zip(actual, expected))
    print(f"  input={values}  expected={expected}  "
          f"current=[{', '.join(show(a) for a in actual)}]  {'ok' if ok else 'DIVERGES'}")

print("dim=-1")
for rows in ([[2.0, 3.0, 4.0], [1.0, 5.0, 3.0]], [[2.0, 0.0, 4.0], [1.0, 5.0, 3.0]]):
    actual, expected = source_prod_bw_dim(rows, 1.0), torch_prod_bw_dim(rows, 1.0)
    ok = all(a == e for ar, er in zip(actual, expected) for a, e in zip(ar, er))
    print(f"  input={rows}  expected={expected}  "
          f"current=[{', '.join('[' + ', '.join(show(a) for a in r) + ']' for r in actual)}]  "
          f"{'ok' if ok else 'DIVERGES'}")
```

Output:

```text
dim=None
  input=[2.0, 3.0, 4.0]  expected=[12.0, 8.0, 6.0]  current=[12, 8, 6]  ok
  input=[2.0, 0.0, 4.0]  expected=[0.0, 8.0, 0.0]  current=[0, non-finite, 0]  DIVERGES
  input=[0.0, 0.0, 4.0]  expected=[0.0, 0.0, 0.0]  current=[non-finite, non-finite, 0]  DIVERGES
  input=[1.0, 5.0, 0.0]  expected=[0.0, 0.0, 5.0]  current=[0, 0, non-finite]  DIVERGES
dim=-1
  input=[[2.0, 3.0, 4.0], [1.0, 5.0, 3.0]]  expected=[[12.0, 8.0, 6.0], [15.0, 3.0, 5.0]]  current=[[12, 8, 6], [15, 3, 5]]  ok
  input=[[2.0, 0.0, 4.0], [1.0, 5.0, 3.0]]  expected=[[0.0, 8.0, 0.0], [15.0, 3.0, 5.0]]  current=[[0, non-finite, 0], [15, 3, 5]]  DIVERGES
```

## Expected behaviour

The gradient of a product with respect to one factor is the product of the *other* factors, which
is always finite for finite inputs:

- no zero in the reduced input: the current formula is correct;
- exactly one zero: zero everywhere except at that position, which gets `grad` times the product
  of the non-zero values;
- two or more zeros: zero everywhere.

Suggested tests: exact zeros in both paths, one zero and several zeros, and a non-finite check
before PCC — a PCC comparison alone can pass or silently degrade on a non-finite element.

**Comment by kanapitsas** [comment]
I have a fix ready for both the dim=None and per-dimension paths, along with a reproduction script. I'm holding the PR until assignment. Could you assign this issue to me?

**Comment by jberkowitzTT** [comment]
## 📝 Background

This is a community-suggested bounty that we have reviewed & assigned to @kanapitsas 


## 💡 Problem to Solve

>ttnn.prod_bw returns non-finite gradients for zero inputs


Note: This bounty is beholden to the [Tenstorrent Bounty Program Terms & Conditions](https://docs.tenstorrent.com/bounty_terms.html) & local tax requirements may apply

**Comment by pieromachacuaycabrera-cyber** [comment]
I have mathematically analyzed the non-finite gradients issue for `ttnn.prod_bw`. The root cause is the naive `prod * grad / input` which evaluates `inf * 0` at zero locations.

To solve this efficiently on device without falling back to per-element branching (which destroys kernel performance), you can use a pure tensor-algebraic masking strategy composed of existing `ttnn` primitive operations.

### The Tensor-Math Solution
Instead of dividing by zero, we compute the product of all *non-zero* elements by forcing zeros to 1, count the zeros, and use `ttnn::where` to select the correct gradient based on the zero-count.

Here is the composite operation flow you need to implement in `unary_backward.cpp`:

```cpp
// 1. Create a mask of exactly where the zeros are
auto is_zero = ttnn::eq(input, 0.0f);

// 2. Create a "safe" input where all 0s are replaced by 1s (so they don't destroy the product)
auto safe_input = ttnn::where(is_zero, 1.0f, input);

// 3. Compute the safe product (product of all non-zero elements)
auto prod_without_zeros = ttnn::prod(safe_input, dim);

// 4. Count total zeros in the reduction dimension
auto zero_count = ttnn::sum(is_zero, dim);

// 5. Build the three gradient conditions:
// Case 0: No zeros -> prod_without_zeros * grad / input  (or standard prod*grad/input)
auto grad_no_zeros = ttnn::multiply(ttnn::multiply(prod_without_zeros, grad), ttnn::reciprocal(input));

// Case 1: Exactly 1 zero -> The gradient is 0 everywhere EXCEPT at the zero's index, where it is (prod_without_zeros * grad)
auto grad_one_zero = ttnn::where(is_zero, ttnn::multiply(prod_without_zeros, grad), 0.0f);

// Case 2: >= 2 zeros -> The gradient is completely 0 everywhere.
auto grad_multi_zero = ttnn::zeros_like(input);

// 6. Select the final gradient using nested where/eq based on zero_count
auto final_grad = ttnn::where(
    ttnn::eq(zero_count, 0.0f),
    grad_no_zeros,
    ttnn::where(ttnn::eq(zero_count, 1.0f), grad_one_zero, grad_multi_zero)
);
```

### Why this fixes the issue:
1. `safe_input` ensures we never divide by zero and never lose the product to a `0 * x` absorption.
2. The nested `ttnn::where` handles the 0, 1, and 2+ zeros cases flawlessly and vectorizes perfectly on the Tenstorrent silicon without divergence.

---
*I hope this patch saves you some time! If you found my contribution helpful, I’d really appreciate a coffee/tip to support my open-source work: https://www.paypal.me/Asi150403*

**Comment by RosMengHeang** [comment]
Hi @jberkowitzTT — interested in #54551 if @kanapitsas is no longer active on it.

**Root cause:** both `dim=None` and per-dim `prod_bw` paths do `reciprocal(input) * broadcast(prod * grad)`. At exact zeros that becomes non-finite (`inf * 0`) instead of the finite autograd value (product of the *other* factors).

**Proposed fix:** zero-aware path —
- 0 zeros → keep current formula  
- exactly 1 zero → grad only at that index = (prod of non-zeros) × grad; elsewhere 0  
- ≥2 zeros → all-zero grad  

Implement with existing `eq` / `where` / `prod` / `sum` primitives (no per-element host fallback). Add finite checks *before* PCC for one-zero and multi-zero cases on both branches.

Can take assignment under the bounty terms and open a clean PR.

**Comment by dressedinblack5** [comment]
Hi — if #54551 opens up for reassignment, I'd like it. Happy to pick this up under the Tenstorrent bounty terms if @kanapitsas is done or unavailable.

**Comment by jberkowitzTT** [comment]
Hi @kanapitsas - congratulations on completing this bounty, and thank you for the great work!

To initiate the payment process, please email us at bounties@tenstorrent.com with a link to this issue
